core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -39,20 +39,28 @@
 import io.trino.sql.planner.plan.PlanNode;
 import io.trino.sql.planner.planprinter.PlanPrinter;
 
+import java.util.Comparator;
+import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Optional;
 import java.util.Set;
 import java.util.function.Predicate;
 import java.util.stream.Stream;
 
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
 import static io.trino.matching.Capture.newCapture;
 import static io.trino.spi.StandardErrorCode.OPTIMIZER_TIMEOUT;
 import static java.lang.String.format;
+import static java.lang.System.nanoTime;
 import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
 import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static java.util.stream.Collectors.joining;
 
 public class IterativeOptimizer
```

Comment:
We already measure this in `stats` field. Can we somehow reuse this? Or maybe we should have layers there where one layer is for JMX and other is for exception?

Reply:
`stats` is global, this is local. I don't see potential for reuse here.